crates/typos-cli/src/bin/typos-cli/report.rs

```diff
@@ -280,6 +286,148 @@ impl Report for PrintJson {
     }
 }
 
+#[derive(Debug)]
+pub(crate) struct PrintSarif {
+    results: Mutex<RefCell<Vec<sarif::Result>>>,
+    error: Mutex<RefCell<Vec<String>>>,
+}
+
+impl Default for PrintSarif {
+    fn default() -> Self {
+        Self {
+            results: Mutex::new(RefCell::new(Vec::new())),
+            error: Mutex::new(RefCell::new(Vec::new())),
+        }
+    }
+}
+
+fn error_to_io_error(error: Box<dyn std::error::Error>) -> std::io::Error {
+    std::io::Error::new(std::io::ErrorKind::Other, error.to_string())
+}
+
+impl Report for PrintSarif {
+    fn report(&self, msg: Message<'_>) -> Result<(), std::io::Error> {
+        match &msg {
+            Message::Typo(msg) => {
+                let start = String::from_utf8_lossy(&msg.buffer[0..msg.byte_offset]);
+                let column_start =
+                    unicode_segmentation::UnicodeSegmentation::graphemes(start.as_ref(), true).count() + 1;
+                let column_end =
+                    unicode_segmentation::UnicodeSegmentation::graphemes(msg.typo, true).count() + column_start;
```

Comment:
1. Looking at the spec, we should be reporting code points, not graphemes
2. We need to specify the column unit

> If a SARIF producer processes text artifacts and theRun.results (§3.14.23) is non-empty, the run object SHALL contain a property named columnKind whose value is a string that specifies the unit in which the analysis tool measures columns. If a SARIF producer processes text artifacts and theRun.results is empty, columnKind MAY be present.
> 
> ...
>
> * "unicodeCodePoints": Each Unicode code point (abstract character) is considered to occupy one column. This means that even a character that is represented in UTF-16 by a surrogate pair is considered to occupy one column.

Reply:
I set the columnKind as `unicodeCodePoints`. So the `graphemes().count()` value should represents the code point correctly.

Reply:
Isn't a `char` a code point while a grapheme can be made of multiple multiple `char`s?

Reply:
As https://www.unicode.org/reports/tr29/#Grapheme_Cluster_Boundaries described, `UnicodeSegmentation::graphemes` should mean to be `unicodeCodePoints` in `SARIF`, because they all represent a user-readable character, even though the underlying layer consists of multiple code points.

> SARIF: This means that even a character that is represented in UTF-16 by a surrogate pair is considered to occupy one column.

> Unicode: For example, “G” + grave-accent is a user-perceived character: users think of it as a single character, yet is actually represented by two Unicode code points. These user-perceived characters are approximated by what is called a grapheme cluster, which can be determined programmatically.

Reply:
> SARIF: This means that even a character that is represented in UTF-16 by a surrogate pair is considered to occupy one column.

From my understanding, UTF-16 surrogate pairs are a way to represent code points that can't be represented otherwise.  Grapheme's are more than just surrogate pairs.

Reply:
I switch to `utf16CodeUnits`